truffle/src/com.oracle.truffle.api.strings/src/com/oracle/truffle/api/strings/TSTaintNodes.java

```diff
@@ -0,0 +1,358 @@
+        /**
+         * Checks whether the given array is tainted in the range {@code from - to}.
+         * @param taint taint array
+         * @param from inclusive
+         * @param to exclusive
+         */
+        public abstract boolean execute(Object[] taint, int from, int to);
+
+        @Specialization(guards = "taint != null")
+        static boolean isTaintedNonNull(Object[] taint, int from, int to) {
+            if (taint == null) {
+                return false;
+            }
+//            assert 0 <= from && from <= to && to <= taint.length;
+            // TODO how to handle constraints?
+            return anyNonNull(taint, from, to);
+        }
+
+        @Specialization(guards = "taint == null")
+        static boolean isTaintedNull(Object[] taint, int from, int to) {
+            return false;
+        }
+
+        private static boolean anyNonNull(Object[] taint, int from, int to) {
+            for (int i = from; i < to; i++) {
+                if (taint[i] != null) {
+                    return true;
+                }
+            }
+            return false;
+        }
+
+        static IsSubArrayTaintedNode getUncached() {
+            return TSTaintNodesFactory.IsSubArrayTaintedNodeGen.getUncached();
+        }
+    }
+
+    @GenerateUncached
+    public abstract static class IsArrayTaintedNode extends Node {
+
+        IsArrayTaintedNode() { }
+
+        /**
+         * Checks whether the given array is tainted.
+         * An array is tainted if it is not {@code null} and has at least
+         * one entry which is also not {@code null}.
+         */
+        public abstract boolean execute(Object[] taint);
+
+        @Specialization(guards = "taint != null")
+        static boolean isTaintedNonNull(Object[] taint,
+                                 @Cached IsSubArrayTaintedNode isSubArrayTaintedNode) {
+            return isSubArrayTaintedNode.execute(taint, 0, taint.length);
+        }
+
+        @Specialization(guards = "taint == null")
+        static boolean isTaintedNull(Object[] taint) {
+            return false;
+        }
+
+        static IsArrayTaintedNode getUncached() {
+            return TSTaintNodesFactory.IsArrayTaintedNodeGen.getUncached();
+        }
+    }
+
+    @GenerateUncached
+    public abstract static class IsTaintedNode extends Node {
+
+        IsTaintedNode() { }
+
+        /**
+         * Checks if the given {@link AbstractTruffleString} is tainted.
+         * A {@link AbstractTruffleString} is tainted iff {@link AbstractTruffleString#taint()} is not {@code null}
+         * and contains at least one element, which is also not {@code null}.
+         */
+        public abstract boolean execute(AbstractTruffleString a);
+
+        @SuppressWarnings("unused")
+        @Specialization
+        static boolean isTainted(AbstractTruffleString a,
+                                 @Cached IsArrayTaintedNode isArrayTaintedNode) {
+            return isArrayTaintedNode.execute(a.taint());
```

Comment:
Now the `IsTaintedNode` is doing more, because of `LazyConcat`. I also liked the idea to have 4 nodes for the major 4 operations (`isTainted`, `addTaint`, `getTaint` and `removeTaint`).